jclklib/proxy/connect.cpp

```diff
@@ -0,0 +1,295 @@
+        return;
+    }
+    for(const auto &p : ports) {
+        const auto &port = p.second;
+        if(port.portState == TIME_RECEIVER) {
+            if(llabs(port.master_offset) > threshold) {
+                printf("port %s (%s) offset %10jd under threshold %d",
+                    port.portid.string().c_str(), port.interface.c_str(),
+                    port.master_offset, threshold);
+                return;
+            }
+        }
+    }
+}
+
+static inline bool sendAction(bool local)
+{
+	static int seq = 0;
+    Message *m;
+    if(local)
+        m = &msgu;
+    else
+        m = &msg;
+    MNG_PARSE_ERROR_e err = m->build(buf, bufSize, seq);
+    if(err != MNG_PARSE_ERROR_OK) {
+        fprintf(stderr, "build error %s\n", msgu.err2str_c(err));
+        return false;
+    }
+    bool ret;
+    if(!local)
+        ret = sk->send(buf, m->getMsgLen());
+        
+    if(!ret) {
+        fprintf(stderr, "send failed\n");
+        return false;
+    }
+    seq++;
+    return true;
+}
+static inline bool sendGet(bool local, mng_vals_e id)
+{
+	bool ret;
+    if(local)
+        ret = msgu.setAction(GET, id);
+    else
+        ret = msg.setAction(GET, id);
+    if(!ret) {
+        fprintf(stderr, "Fail get %s\n", msg.mng2str_c(id));
+        return false;
+    }
+    return sendAction(local);
+}
+
+//TODO: subscription part
+bool send_subscription(struct jcl_handle **handle)
+//bool send_subscription()
+{
+    memset(d.bitmask, 0, sizeof d.bitmask);
+    d.duration = SUBSCRIBE_DURATION;
+    d.setEvent(NOTIFY_PORT_STATE);
+    d.setEvent(NOTIFY_TIME_SYNC);
+
+    //TODO: subscription part
+/*   struct jcl_subscription subscription;
+    subscription.event = jcl_offset_threshold_flag | jcl_peer_present_flag;
+	subscription.threshold.offset = 100;
+    handle->subscription = subscription;
+*/
+
+    if(!msg.setAction(SET, SUBSCRIBE_EVENTS_NP, &d)) {
+        fprintf(stderr, "Fail set SUBSCRIBE_EVENTS_NP\n");
+        return false;
+    }
+    bool ret = sendAction(false);
+    msg.clearData(); // Remove referance to local SUBSCRIBE_EVENTS_NP_t
+    return ret;
+}
+
+void *loop( void *arg)
+{
+    const uint64_t timeout_ms = 1000; // 1 second
+
+    for(;;) {
+        if(!sendGet(true, PORT_PROPERTIES_NP))
+            goto end_loop;
+        ssize_t cnt;
+
+        sk->poll(timeout_ms);
+        cnt = sk->rcv(buf, bufSize);
+
+        if(cnt > 0) {
+            MNG_PARSE_ERROR_e err = msg.parse(buf, cnt);
+            if(err == MNG_PARSE_ERROR_OK) {
+                handle(true);
+                break;
+            }
+        }
+    }
+
+    // Send to all
+    sendGet(false, PORT_PROPERTIES_NP);
+
+    while (1) {
+        if (epoll_wait( epd, &epd_event, 1, 100) != -1) {
+        //if ( sk->poll(timeout_ms)) {
+            const auto cnt = sk->rcv(buf, bufSize);
+            MNG_PARSE_ERROR_e err = msg.parse(buf, cnt);
+            if(err == MNG_PARSE_ERROR_OK)
+                handle(false);
+        }
+    }
+end_loop:
+    return;
```

Comment:
connect.cpp: In function 'void* loop(void*)':
connect.cpp:244:5: error: return-statement with no value, in function returning 'void*' [-fpermissive]
  244 |     return;
      |     ^~~~~~